Add `intel-oneapi-runtime`, allow injecting virtual dependencies

alalazo opened the pull request:
This PR adds:
- A new runtime for `%oneapi` compilers, called `intel-oneapi-runtime`
- Information to both `gcc-runtime`  and `intel-oneapi-runtime`, to ensure that we don't mix compilers using different soname for either `libgfortran` or `libifcore`

To do so, the following internal mechanisms have been implemented:
- Possibility to inject virtual dependencies from the `runtime_constraints` callback on packages

Information has been added to `gcc-runtime` to provide the correct soname under different conditions on its `%gcc`.

Rules injected into the solver looks like:

<details>

<summary>Injection of virtual dependency</summary>

```prolog
% Add a dependency on 'gfortran@5' for nodes compiled with gcc@=13.2.0 and using the 'fortran' language
attr("dependency_holds", node(ID, Package), "gfortran", "link") :-
  attr("node", node(ID, Package)),
  attr("node_compiler", node(ID, Package), "gcc"),
  attr("node_compiler_version", node(ID, Package), "gcc", "13.2.0"),
  not external(node(ID, Package)),
  not runtime(Package),
  attr("language", node(ID, Package), "fortran").

attr("virtual_node", node(RuntimeID, "gfortran")) :-
  attr("depends_on", node(ID, Package), ProviderNode, "link"),
  provider(ProviderNode, node(RuntimeID, "gfortran")),
  attr("node", node(ID, Package)),
  attr("node_compiler", node(ID, Package), "gcc"),
  attr("node_compiler_version", node(ID, Package), "gcc", "13.2.0"),
  not external(node(ID, Package)),
  not runtime(Package),
  attr("language", node(ID, Package), "fortran").

attr("node_version_satisfies", node(RuntimeID, "gfortran"), "5") :-
  attr("depends_on", node(ID, Package), ProviderNode, "link"),
  provider(ProviderNode, node(RuntimeID, "gfortran")),
  attr("node", node(ID, Package)),
  attr("node_compiler", node(ID, Package), "gcc"),
  attr("node_compiler_version", node(ID, Package), "gcc", "13.2.0"),
  not external(node(ID, Package)),
  not runtime(Package),
  attr("language", node(ID, Package), "fortran").
```

</details>

alalazo committed:
Add a "language" directive

alalazo committed:
gcc-runtime: model gfortran ABI compatibility

alalazo commented on lib/spack/spack/solver/display.lp:
```diff
@@ -50,3 +50,5 @@
 #show imposed_nodes/3.
 
 % debug
+max_dupes("gfortran", 1).
+possible_in_link_run("gfortran").
```
Don't mind these details, PR still needs some polishing

alalazo committed:
intel-oneapi-runtime: add a new runtime package

The intel-oneapi-runtime depends on gcc-runtime at link time.

This dependency can be used later to dynamically select which GCC to use
with the %oneapi compiler.

intel-oneapi-runtime provides ifort for packages that need fortran

alalazo committed:
Add a fortran-rt virtual for all Fortran runtimes

alalazo commented:
First benchmark as of https://github.com/spack/spack/pull/42062/commits/661f908a4cbc199c93a9d88391bd9ae18b9799cc

* **Spack:** 0.22.0.dev0 (2f935b325f80c159655c757837c218aa7e750760)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

develop: 66813460c0777a4a2bbeb071db3cd7318f382811
PR: 661f908a4cbc199c93a9d88391bd9ae18b9799cc

[radiuss.develop.csv](https://github.com/spack/spack/files/14022480/radiuss.develop.csv)
[radiuss.pr.csv](https://github.com/spack/spack/files/14022481/radiuss.pr.csv)
[radiuss.txt](https://github.com/spack/spack/files/14022482/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/b5ebfb16-9d8a-4b72-a7c0-c07996da23cb)

Adding more rules to the runtimes impacts the solve time

alalazo commented:
On the other hand, with https://github.com/spack/spack/commit/661f908a4cbc199c93a9d88391bd9ae18b9799cc we have:

```yaml
spack:
  specs:
  - hdf5%oneapi+fortran~mpi
  packages:
    cmake:
      require: "%gcc"
  view: true
  concretizer:
    unify: true
```

![hdf5](https://github.com/spack/spack/assets/4199709/f15b8db5-d49f-45e6-97ab-3f10e4136cf0)

Not that in the graph we introduced `intel-oneapi-runtime`, with a dependency on `gcc-runtime` (which in principle allows the selection of the underlying `gcc`).

alalazo committed:
Use a symlink to avoid licensing issues

alalazo committed:
Group addition of dynamically injected virtual packages into a function

alalazo commented:
@spackbot run pipeline

alalazo commented:
@spackbot run pipeline

haampie commented on var/spack/repos/builtin/packages/gcc-runtime/package.py:
```diff
@@ -44,9 +47,15 @@ class GccRuntime(Package):
     def install(self, spec, prefix):
         if spec.platform in ["linux", "cray", "freebsd"]:
-            libraries = self._get_libraries_elf()
+            libraries = get_elf_libraries(compiler=self.compiler, libraries=self.LIBRARIES)
```
what's this change about?

alalazo replied:
This is just to reuse the same code in `intel-oneapi-runtime`. It's temporary, and I think in the end we should probably move this somewhere in core Spack (my guess is that it will be used by most runtimes).

haampie commented on var/spack/repos/builtin/packages/gcc-runtime/package.py:
```diff
@@ -44,9 +47,15 @@ class GccRuntime(Package):
     ]
 
+    # libgfortran ABI
+    provides("fortran-rt", "gfortran")
```
drop the latter?

alalazo replied:
So, the idea is that `fortran-rt` will prevent using two different runtimes in the same unification set, while `gfortran` will encode soname compatibility. Does that make sense to you?

haampie commented on etc/spack/defaults/packages.yaml:
```diff
@@ -24,12 +24,15 @@ packages:
       elf: [elfutils]
       fftw-api: [fftw, amdfftw]
       flame: [libflame, amdlibflame]
+      fortran-rt: [gcc-runtime, intel-oneapi-runtime]
```
Is the consequence of this that every compiler for fortran must provide a separate runtime package that provides it? Are there cases where there's basically nothing to provide at all? Wondering if it shouldn't be on the compiler itself (can compile fortran vs provides fortran libraries)

alalazo replied:
> Is the consequence of this that every compiler for fortran must provide a separate runtime package that provides it? 

Currently it's like that, but nothing prevents not to use the runtimes. Ultimately it's up to the compilers to decide which rules apply when they get used. If we have a compiler that doesn't need a separate runtime, we can encode that.

To be clear, right now it's the compilers that say "each package needing fortran and compiling with me needs `fortran-rt`, and I am the provider of it".

tgamblin replied:
I'm not seeing why the preference here would mandate a compiler to have a runtime class -- seems up to how we want to model things, so I agree w/Max.

haampie commented on var/spack/repos/builtin/packages/intel-oneapi-runtime/package.py:
```diff
@@ -0,0 +1,52 @@
+    provides("fortran-rt", "ifcore@5", when="%oneapi@2021:")
+
+    def install(self, spec, prefix):
+        if spec.platform in ["linux", "cray", "freebsd"]:
```
requires?

mabraham commented:
@rscohn2 Of possible interest to you

rscohn2 commented on var/spack/repos/builtin/packages/intel-oneapi-runtime/package.py:
```diff
@@ -0,0 +1,52 @@
+        "ifcore",  # Fortran
+        "ifcoremt",  # Fortran
+        "ifport",  # Fortran
+        "sycl",
```
What about libiomp5?

I have a problem now where mkl wants to depend on a virtual openmp runtime that could be provided by gcc or icx: https://github.com/spack/spack/issues/42310

rscohn2 commented on var/spack/repos/builtin/packages/intel-oneapi-runtime/package.py:
```diff
@@ -0,0 +1,52 @@
+            tty.warn("Could not detect any shared OneAPI runtime libraries")
+            return
+
+        for path, name in libraries:
```
Instead of doing the symlink, the oneapi documentation tells people to make the prefix 2 levels up: https://spack.readthedocs.io/en/latest/build_systems/inteloneapipackage.html#libraries
and the packages know how to find the bin/lib/include 2 levels down. I tried introducing symlinks a while ago, but got rid of them for 2024 because of some conflicts (I only remember that it was MPI, not the details). If I was starting from nothing, I would introduce the symlinks on day 1.

This is a new package, so symlinks seems ok, and people will have to know that the external package prefix is not 2 levels up.

alalazo replied:
@rscohn2 For other packages, like `gcc-runtime` we would actually copy the libraries in place. I didn't do it here because I think it would be a license infringement for OneAPI (if e.g. nodes are redistributed via `buildcache push`), whereas a symlink would not be an issue. Is that correct, or do you have any better suggestion?

rscohn2 replied:
The library runtimes have a separate license and allow redistribution.

haampie replied:
nice

alalazo replied:
With respect to https://github.com/spack/spack/pull/42062#issuecomment-1999796776

@rscohn2 That's what we were doing initially, before this comment of yours :point_up: I'm a bit confused at the moment on what we can do... :sweat_smile:

rscohn2 replied:
Todd suggested checking the redist to make sure the files were listed. When I did that, I saw that redistribution has obligations, like disallowing reverse engineering. It is ok to symlink, it is ok to copy, but it would be complicated to redistribute.  I had forgotten about the discussion above, but now I see that my answer was not a good one. Sorry if this caused you to do unnecessary work. Do you think there is a benefit to going back to symlinks vs marking the package as non-cacheable?

tgamblin replied:
> I saw that redistribution has obligations, like disallowing reverse engineering

@rscohn2: can you say why this complicates redistribution?  Seems like redistribution is allowed, and we expect users to comply with the licenses for packages. What's the extra complication here (other than another condition of the license)?

rscohn2 replied:
This is the scenario the redistribution license was intended to support:

I build an application with icpx. I distribute the application with some intel libraries (e.g. libsycl, libimf, ...) I attach a license to my distribution. I need to make sure my use of the libraries complies with intel's license. 

This includes restricting the rights of people who use my distribution (e.g. no reverse engineering of intel libraries). My distribution needs to have a license that passes on the intel restrictions. This is doable. But it seems better to disable redistribution of the package.

tgamblin replied:
One of our goals is to allow users to easily build containers that use the runtime components of the intel compilers, and to keep the containers reasonably sized.  I think we're going to have a much harder time with the second part of that if we have to bundle the entire compiler vs. just the runtime.  So I think the goal here is to support the use case as described.

I'm not sure how redistributing vs. not redistributing makes a difference in users' ability to reverse engineer the libraries. They could download these libraries from intel if they wanted to reverse engineer, right?  The only difference is that they'd be getting the runtime libraries in a spack package.  If we need to talk to Intel lawyers to get on the same page with them we could do that -- e.g., UO has an arrangement that allows them to redistribute the entire intel compiler package (which we do not -- we don't put the intel packages in our mirror).

rscohn2 replied:
It is possible to redistribute the libraries. Intel has specifically created a license that lets you do that. Someone has to do due diligence to ensure that the way the libraries are packaged and distributed complies with the license. That has not happened. Even a BSD license comes with obligations when it comes to redistribution. If HPSF will have a working group for package/distribution, then it would be good to cover licenses there.

alalazo committed:
intel-oneapi-runtime: copy libraries, instead of symlonking

Also: provides sycl, add libiomp5 to the list of libraries

alalazo commented:
I have been able to compile `gromacs` using OneAPI `sycl` implementation by using the following environment:
```yaml
spack:
  specs:
  - gromacs+sycl +intel_provided_gcc %oneapi
  packages:
    cmake:
      require:
      - "%gcc"
    sycl:
      require:
      - "intel-oneapi-runtime"
    fftw-api:
      require:
      - "intel-oneapi-mkl"
    lapack:
      require:
      - "intel-oneapi-mkl"

  view: true
  concretizer:
    unify: true
```
(and a few hacks in `compilers.yaml` because some Intel-OneAPI executables on Ubuntu have wrong runpath :sweat: )

rscohn2 commented:
> (and a few hacks in `compilers.yaml` because some Intel-OneAPI executables on Ubuntu have wrong runpath 😓 )

I reported it and am hoping it will be fixed in the next release.

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2433,14 +2446,23 @@ def visit(node):
 
+    def add_runtime_virtuals(self):
+        """Explicitly add the virtuals that might be attached dynamically by runtimes"""
+        for vpkg in ("fortran-rt", "ifcore", "gfortran"):
```
I think we're going to need a `runtimes` module at some point. Could we move this list there instead of having it ad-hoc in the solver?

alalazo replied:
I can probably iterate over packages tagged as "runtimes", and check which virtuals they could provide. Let me try that.

tgamblin replied:
Why aren't these added based on what is passed to `depends_on` in `RuntimePropertyRecorder`?

You know that someone executed `runtime_constraints` and added a possible dependency on these virtuals, and you know that they're virtual because they're provided by the runtime packages.  This shouldn't need to be hard-coded...

alalazo replied:
I'll check if it is easy to unify that. This was added here to follow the same code path for max duplicate counters as for everything else, but it makes sense to aggregate the way you say.

alalazo replied:
I removed this in https://github.com/spack/spack/pull/42062/commits/b9c497abec6db0f0665fc7f6a454823bc961bf7b and moved everything to the counters

tgamblin replied:
I don't think moving this to the counters really solves the issue.  That replaces the list with tags that users have to add.  See my comment below (and the one above) -- we should just *know* what possible virtuals there are before solve.  You do not know them from `possible_dependencies()`, but you *do* know them once you've gone through those dependencies, figured out which ones have a `runtime_constraints()` method, and called it for each of them.

becker33 commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -377,3 +377,33 @@ def _ld_library_path(self):
+                f"{str(compiler.spec)} and using the 'fortran' language",
+            )
+        # The version of gcc-runtime is the same as the %gcc used to "compile" it
+        pkg("intel-oneapi-runtime").requires(
```
Why isn't this covered by the version constraint on the `intel-oneapi-runtime` dependency above?

alalazo replied:
The constraint above fixes the minimum version of `intel-one-api` runtime given the compiler version used to compile a package X depending on it. This one relates the version of `intel-oneapi-runtime` to the version of the oneapi compiler "attached" to it. It's the same pattern used for gcc-runtime.

becker33 commented on var/spack/repos/builtin/packages/intel-oneapi-runtime/package.py:
```diff
@@ -0,0 +1,54 @@
+    def install(self, spec, prefix):
+        if spec.platform in ["linux", "cray", "freebsd"]:
+            libraries = get_elf_libraries(compiler=self.compiler, libraries=self.LIBRARIES)
+        else:
+            raise RuntimeError("Unsupported platform")
```
Presumably this requires some future work for macos and windows? I thought oneapi supports all of them

alalazo replied:
Windows is supported (but not accounted for in this PR). macOS is unsupported by OneAPI, as far as I can tell.

becker33 commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -1155,5 +1155,21 @@ def runtime_constraints(cls, *, compiler, pkg):
             f"it depends on gcc-runtime@{str(compiler.version)}:",
         )
+
+        gfortran_str = "gfortran@5"
+        if compiler.spec.satisfies("gcc@:6"):
+            gfortran_str = "gfortran@3"
+        elif compiler.spec.satisfies("gcc@7"):
+            gfortran_str = "gfortran@4"
```
This seems a bit hacky, trying to make sure I've wrapped my head around it.

The alternative here seems to be separate runtime nodes for `gcc-fortran-runtime`, `gcc-c-runtime`, `gcc-cxx-runtime` and having those nodes versioned in a manner that includes the fortran runtime versioning?

That doesn't seem like it's better, but just trying to make sure I understand what's going on here.

alalazo replied:
Yes. `gfortran` is a virtual, and it models the soname of the `libgfortran` library. Depending on the version of the compiler, the package provides the virtual at different versions - which should model soname incompatibilities.

tgamblin replied:
I think the virtual for this and other models of library sonames should really be prefixed with `lib` (i.e., this should be called `libgfortran`) if we're modeling the soname so closely.  `lib` can indicate that we really mean the specific library.

alalazo replied:
Done in https://github.com/spack/spack/pull/42062/commits/f823ed45a22e3b88cd5bc0b8433d57ef1cf7ce40

alalazo replied:
I also changed `ifcore` -> `libifcore`

tgamblin commented on var/spack/repos/builtin/packages/intel-oneapi-runtime/package.py:
```diff
@@ -0,0 +1,54 @@
+
+    depends_on("gcc-runtime", type="link")
+
+    LIBRARIES = [
+        "imf",
+        "intlc",
+        "irng",
+        "svml",
+        "ifcore",  # Fortran
+        "ifcoremt",  # Fortran
+        "ifport",  # Fortran
+        "iomp5",
+        "sycl",
```
We should have some kind of check to ensure that all the libraries listed here in `redist.txt` or `resist-rt.txt` in the compiler distribution, since that's where it's documented authoritatively. See https://www.intel.com/content/www/us/en/developer/articles/license/end-user-license-agreement.html (oneapi license part) for details -- the PDF there outlines what we're allowed to do.d
I *think* we're allowed to have these in their own package though.

alalazo replied:
Not clear to me where to look :worried: 
```console
$ find /opt/intel/oneapi/compiler -name redist.txt 
$ find /opt/intel/oneapi/compiler -name *.txt | wc -l
81
```

alalazo replied:
@rscohn2 Do you have any idea where I need to look?

rscohn2 replied:
These files are redistributable, but redistribution carries some obligations and this is not a case they anticipated. Redistribution usually means giving to another legal entity. It looks like you are just copying the files to a different directory? If that is all you are doing then I don't think there are license issues.

alalazo replied:
Sure, @rscohn2 The point is though that those other directories can then be packaged in a buildcache and redistributed via a public mirror, for instance from our `oneapi` pipeline. Would that be ok?

rscohn2 replied:
I think the intention for the redistribution license was that you are
distributing an application that needs the libraries. This looks like
something different. Rather than try to get lawyers involved, could you
mark this package as non cacheable? That is what I did for all the
intel-oneapi-* packages. I recognize that it is less convenient.

https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/oneapi.py#L30

alalazo replied:
@rscohn2 

> https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/oneapi.py#L30

I was not aware of this class level attribute. Do you know how it is used? From what I can tell:
```console
$ grep redistribute_source -R lib
lib/spack/spack/build_systems/oneapi.py:    redistribute_source = False
```
it is not used directly in Spack?

rscohn2 replied:
I did not know it was never implemented: https://github.com/spack/spack/pull/20185

I suggest you keep the copying code as it is now. It seems more robust and consistent with gcc implementation. I will look at implementing the feature that prevents packages from being published. Is @scheibelp the right person to ask about it.

alalazo replied:
@rscohn2 Thanks for the reply. I think @scheibelp is the correct person to continue working on #20185

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2433,14 +2446,23 @@ def visit(node):
         recorder = RuntimePropertyRecorder(self)
         for compiler in self.possible_compilers:
-            if compiler.name != "gcc":
+            if compiler.name not in ("gcc", "oneapi"):
```
We shouldn't special case things here. 

First thought: this should look at whether the compiler's corresponding package class has a `runtime_constraints()` method and call it if it does. 

Really, I don't think there should be anything special about compilers -- *any* class could have a `runtime_constraints` method.  The only reason it's special for compilers right now is because `runtime_constriants(compiler, pkg)` has to take a compiler spec as a parameter, and that is only because we're in transition from compilers to compilers as proper packages.

it would be nice if in this PR we could get rid of the special casing.  Is there any reason why we can't convert this:

```python
    @classmethod
    def runtime_constraints(cls, compiler, pkg):
        ...
```

to something like:

```python
    @classmethod
    def global_constraints(cls, spec, pkg):
        ...
```

And make the spec an actual *spec* (not compiler spec) for the compiler's associated package? I expect it will need to be this eventually, anyway.  I don't think this is a huge lift, because all we're using the compiler spec for is to get a version and/or name string.  Full fledged specs have those things too.

I think starting with at least that would be good, and eventually I think we want this method to loop over all the possible dependencies and allow them to define global constraints if they need to.  I care about this because I think this pattern has use cases beyond just compilers, and I want the logic to be general.

alalazo replied:
> We shouldn't special case things here.

This specific check is not needed anymore, and will be removed. 

For the API, I think we can inject a spec, but we might need to attach some more information to it. Currently we are not passing a `CompilerSpec`, but a `Compiler`. This means we would be missing information on `cc`, `cxx`, etc. paths (in case they are needed).

alalazo replied:
Opted for passing a `CompilerSpec` for the time being, to avoid double conversions like `oneapi` -> `intel-oneapi-compilers` before the calls, and the opposite within the call.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2433,14 +2446,23 @@ def visit(node):
+            if compiler.name not in ("gcc", "oneapi"):
                 continue
+
+            compiler_with_different_cls_names = {"oneapi": "intel-oneapi-compilers"}
```
This mapping should live in one place ... ideally not here. Don't we already have it somewhere for bootstrapping compilers the old way?  In any case, this and `possible_compilers` should be the only special cases IMO.

becker33 replied:
We have this in `lib/spack/spack/compilers/__init__.py`.

tgamblin replied:
This still needs to be consolidated - it should come from `compilers/__init__.py`

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2856,13 +2878,22 @@ def reset(self):
             dependency_str: the dependency spec to inject
             when: anonymous condition to be met on a package to have the dependency
             type: dependency type
+            languages: languages needed by the package for the dependency to be needed
```
needs rewording -- I dunno what this means

alalazo replied:
Clearly a rebasing issue or something. See https://github.com/spack/spack/pull/42062/commits/b0e97fb761ebf8e7f0f5b725538bf31b074803d6

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -158,6 +158,14 @@ error(100, multiple_values_error, Attribute, Package)
+% Languages used
+%-----------------------------------------------------------------------------
+
+attr("language", node(X, Package), Language) :-
+  condition_holds(ConditionID, node(X, Package)),
+  pkg_fact(Package,language(ConditionID, Language)).
```
Is the only reason we're not modeling a language as a dependency basically because the language dependency doesn't have a well defined deptype?  I still don't see much of a difference between this and a dependency condition for a virtual.

alalazo replied:
See https://github.com/spack/spack/pull/42062#discussion_r1521135019

tgamblin replied:
This should eventually be unnecessary once we have compilers modeled as packages.

tgamblin commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -1155,5 +1155,21 @@ def runtime_constraints(cls, *, compiler, pkg):
+        for fortran_virtual in ("fortran-rt", gfortran_str):
+            pkg("*").depends_on(
+                fortran_virtual,
+                when=f"%{str(compiler.spec)}",
```
What is this going to become when we no longer have the compiler attribute on the node?  How will we test whether something was "compiled" with `gcc` then?

this is a question for @becker33 too.

alalazo replied:
When compilers are dependencies, I think this will be `when="^[virtual=fortran] gcc@X.Y"`

tgamblin commented on var/spack/repos/builtin/packages/hdf5/package.py:
```diff
@@ -33,6 +33,10 @@ class Hdf5(CMakePackage):
 
     license("custom")
 
+    language("c")
+    language("cxx", when="+cxx")
+    language("fortran", when="+fortran")
```
I like this a lot but I think it would be even clearer (and less cognitive load for users) as:

```python
    depends_on("c")
    depends_on("cxx", when="+cxx")
    depends_on("fortran", when="+fortran")
```

I'm having a hard time understanding why we need `language()` when we could do this.

alalazo replied:
In my opinion we need `language`[^1] because it decouples the information we want to add on language support from the `depends_on` semantic. This has two effects:
1. We can proceed improving runtimes even without having compilers turned into dependencies
2. Later, we have a place where we can extend or change the implementation if need be

To explain point 1, if we used:
```console
depends_on("fortran", when="+fortran")
```
and assumed `fortran` is a virtual, then we would need a provider for that virtual. The provider in the final software model is a compiler, treated as a dependency. This means runtime improvements introducing some information on language support are **blocked** on having compilers as proper dependencies. If we have a directive like `language`, we can adopt an implementation now that fits compilers as node attributes, and change it later to fit compilers as dependencies.[^2]

For point 2, I agree that when we turn compilers into dependencies, we would likely have:
```python
def language(lang_str, when=when_str):
    return depends_on(lang_str, type="build", when=when_str)
```
as a first implementation. The difference with having `depends_on` directly would be that here we have a place where we can decorate or change the implementation if need be. Basically, we differentiate "runtimes" from the start and we have a way to selectively modifying them without modifying packages .

[^1]: By `language` I mean an intermediate directive, I'm not focusing on its name, which can be changed if we have a better one.

[^2]: It's my impression that improvements on runtimes are at the moment much more reachable than compilers as deps, for which we still have some design points to be decided, so this is in my opinion relevant because otherwise we might introduce significant delays.

tgamblin replied:
I specifically do not want to differentiate runtimes. They’re dependencies like anything else, and should be modeled that way. This is fundamental to having a general model. I don’t want spack to be like other systems where toolchains are special — that is what we are trying to get away from.

I don’t see why we can’t implement the differentiation *under* depends_on and eventually remove it.

as it is now, a user has to understand why language is different from depends_on. They shouldn’t have to, especially if we do not intend for there to be a difference.

tgamblin replied:
I to elaborate on this a little more — I don’t think “languages” are the only things that will need the extra semantics. I think build tools, instrumentation tools, profilers, sanitizers, etc. are going to want to do similar things, and they’re not “languages”. So o think it doesn’t make sense to narrow the meaning of the directive and special case the functionality when it’s really something an arbitrary tool could end up using. It makes more sense to me to keep it general with depends_on.

alalazo replied:
> I specifically do not want to differentiate runtimes.

The runtimes are not differentiated. There's a difference between the language (which brings in the compiler) and the runtimes that are _eventually_ injected by the compiler. 

I can technically do the opposite right now (pseudo-code below):
```python
def depends_on(spec_str, type=("build", "link"), when=when_str):
    if spec_str in SUPPORTED_LANGUAGES:
        # Different treatment of languages until compilers are deps
        assert type == "build"
        return _language(spec_str, type="build", when=when_str)
    
    # Usual code path for other deps
```
but note that this will force us to use the `depends_on` API later. If we need additional arguments, or anything similar, to deal with languages at a later stage - and that is not 100% clear to me yet - we'll have to change this API.

alalazo replied:
> I think build tools, instrumentation tools, profilers, sanitizers, etc. are going to want to do similar things, and they’re not “languages”.

Footnote above: "By language I mean an intermediate directive, I'm not focusing on its name, which can be changed if we have a better one."

> I think build tools, instrumentation tools, profilers, sanitizers, etc. are going to want to do similar things, and they’re not “languages”.

Do we have use cases for these? At the moment I can't think of anything else than tools that in some way wrap compilers - and so they can probably be considered compilers too? I'm thinking about tools like `tau` or `scalasca` etc.

alalazo replied:
Done in https://github.com/spack/spack/pull/42062/commits/3f930e27a8cae2775feb922b73e414c67ab160de

alalazo committed:
intel-oneapi-runtime: replace install time check by directives

alalazo committed:
Renamed gfortran -> libgfortran

alalazo committed:
Don't hardcode names when counting possible packages

alalazo committed:
Remove the "language" directive, in favor of "depends_on"

alalazo committed:
Fix wording in docstring

alalazo committed:
Remove needless check on supported compilers

alalazo committed:
gcc-runtime: fix issue introduced with renaming

alalazo committed:
intel-oneapi-runtime: add stub libs, as for gcc-runtime

alalazo committed:
Renamed ifcore -> libifcore

alalazo committed:
Pass a Spec instead of a Compiler

For now maintain a CompilerSpec, since it allows to
avoid a double conversion from compiler name to pkg
name and vice-versa

rscohn2 commented:
I think the intention for the redistribution license was that you are
distributing an application that needs the libraries. This looks like
something different. Rather than try to get lawyers involved, could you
mark this package as non cacheable? That is what I did for all the
intel-oneapi-* packages. I recognize that it is less convenient.

On Fri, Mar 15, 2024 at 10:22 AM Massimiliano Culpo <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/intel-oneapi-runtime/package.py
> <https://github.com/spack/spack/pull/42062#discussion_r1526367070>:
>
> > +    LIBRARIES = [
> +        "imf",
> +        "intlc",
> +        "irng",
> +        "svml",
> +        "ifcore",  # Fortran
> +        "ifcoremt",  # Fortran
> +        "ifport",  # Fortran
> +        "iomp5",
> +        "sycl",
>
> Sure, @rscohn2 <https://github.com/rscohn2> The point is though that
> those other directories can then be packaged in a buildcache and
> redistributed via a public mirror, for instance from our oneapi pipeline.
> Would that be ok?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/42062#discussion_r1526367070>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOY2WU6FL3CSQ7V5BYPSBDYYL73VAVCNFSM6AAAAABBX3UBHGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMZZGA4DQNRXG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -585,6 +588,9 @@ def depends_on(
 
     """
+    if spack.spec.Spec(spec).name in SUPPORTED_LANGUAGES:
+        assert type == "build", "languages must be of 'build' type"
```
Why is this needed?  It's not really a build dependency; it's a build dependency on the compiler *and* (maybe) a link dependency on the runtime, depending on what the compiler decides to do. 

Languages should just omit the deptype. In practice, it doesn't matter b/c it's determined by the compiler's package (and some may not need a runtime). The caller of `depends_on` should not have to do anything based on *how* the language is implemented. they should just say they depend on it.

alalazo replied:
> Why is this needed? It's not really a build dependency; **it's a build dependency on the compiler** and (maybe) a link dependency on the runtime, depending on what the compiler decides to do.

Depending on a language just brings in the compiler, so as you say above it's a build dependency on the compiler.  Then the compiler may or may not decide to inject further dependencies. If we don't make the language dependency a build dependency we won't be able to have duplicates (since link dependency virtuals are unified, whereas build are not).

> Languages should just omit the deptype.

But `depends_on` have that argument, so a user may specify it. That was one of the reason for which I started implementing this using another directive. If we use `depends_on` I think people need to get used to adding `type="build"` as they do e.g. for `cmake`. If we use a wrapper directive we can take control of the type, and avoid exposing it.

tgamblin replied:
> I think people need to get used to adding type="build" as they do e.g. for cmake.  If we use a wrapper directive we can take control of the type, and avoid exposing it.

You can just do that with `depends_on`.  No extra directive, no extra convention. Don't have them specify the deptype.  It's the "default" deptype, and the machinery underneath should take care of the rest.

tgamblin commented on lib/spack/spack/package_base.py:
```diff
@@ -567,6 +567,7 @@ class PackageBase(WindowsRPath, PackageViewMixin, metaclass=PackageMeta):
     provided_together: Dict["spack.spec.Spec", List[Set[str]]]
     patches: Dict["spack.spec.Spec", List["spack.patch.Patch"]]
     variants: Dict[str, Tuple["spack.variant.Variant", "spack.spec.Spec"]]
+    languages: Dict["spack.spec.Spec", Set[str]]
```
This should be TODO to remove later.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2889,26 +2910,45 @@ def depends_on(self, dependency_str: str, *, when: str, type: str, description:
             f"  not external({node_variable}),\n"
             f"  not runtime(Package)"
         ).replace(f'"{placeholder}"', f"{node_variable}")
+        if languages:
+            body_str += ",\n"
+            for language in languages:
+                body_str += f'  attr("language", {node_variable}, "{language}")'
```
this, like `depends_on()` should be using `SUPPORTED_LANGUAGES` an not forcing extra parameters on the caller. 

TODO: the final implementation of this should not have an extra `languages` parameter and should not have any special cases for "languages".  There should be one general mechanism for packages to inject additional dependencies, and that's all we should need.

tgamblin commented on lib/spack/spack/solver/counter.py:
```diff
@@ -24,7 +25,13 @@ class Counter:
 
     def __init__(self, specs: List["spack.spec.Spec"], tests: bool) -> None:
-        self.specs = specs
+        runtime_pkgs = spack.repo.PATH.packages_with_tags("runtime")
+        runtime_virtuals = set()
+        for x in runtime_pkgs:
+            pkg_class = spack.repo.PATH.get_pkg_class(x)
+            runtime_virtuals.update(pkg_class.provided_virtual_names())
+
+        self.specs = specs + [spack.spec.Spec(x) for x in runtime_pkgs]
```
This requires too much effort from the packager. They shouldn't have to tag runtimes.

I'm assuming the reason we do this is because we need the full set of possible virtuals before the solve, but you *have* that after you call `runtime_constraints()` on all the packages that have that method. You can just see if any of the global dependencies added with `pkg("*").depends_on()` are virtual, and get the same (actually, possibly smaller) list of `runtime_virtuals` that you have here.  So this logic should go away and we should just look at what was injected.

alalazo replied:
I can remove the tags for an extra method in e.g. `gcc` that returns what the package could possibly inject.

tgamblin commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -1149,11 +1149,27 @@ def runtime_constraints(cls, *, compiler, pkg):
+            gfortran_str = "libgfortran@4"
+
+        for fortran_virtual in ("fortran-rt", gfortran_str):
+            pkg("*").depends_on(
```
This `depends_on()` is the one that should alert the solver to the fact that `libgfortran` and `fortran-rt` are possible virtuals.  I think the `pkg` object should keep track of that and we should use it once all `runtime_constraints()` have been called in setup.

tgamblin commented on var/spack/repos/builtin/packages/hdf5/package.py:
```diff
@@ -33,6 +33,9 @@ class Hdf5(CMakePackage):
 
     license("custom")
 
+    depends_on("cxx", type="build", when="+cxx")
+    depends_on("fortran", type="build", when="+fortran")
```
@alalazo Why do these have to be build dependencies? I think a default deptype is fine here, as this will eventually resolve to a build dependency on a compiler and link dependencies on libraries.

alalazo replied:
Discussed in DMs. We might want to consider a new deptype. Currently the dependency maps to the compiler only, which needs to be `build` due to unification rules (once compilers are real dependencies).

tgamblin reviewed (commented):
@alalazo I'm merging this as it is far enough along that I think we can merge without hurting anything, and it can be improved once compilers are full packages.

I am still not sure that the interface is 100% yet. Specifically, so we don't forget:

1. The deptype of a `depends_on()` on a language either shouldn't matter or should just be default. The deptype(s) is (are) *entirely* up to the compiler package... not the caller, so I don't see any point in forcing an arbitrary convention. Omitting is better.
2. `pkg("*").depends_on()` shouldn't take a language parameter and we should not make languages special. They're just virtuals like any other... all we're adding is a means for dependencies to inject *additional* dependencies on their dependents. There is some fanciness I think we will need to do in solver setup to enable this but we should not force callers to know these special cases -- it should just work.
3. I think likewise we shouldn't need to special case languages in the solve, at least not after compilers are proper dependencies. That can change later, too.

More comments/details below, but this is looking better and is moving along!

balay commented:
I'm seeing the following behavior with this PR change.
```
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack.a$ ./bin/spack spec xsdk@1.0.0 %oneapi@2024.0.2 |grep hdf5@
 -           ^hdf5@1.14.3%gcc@11.4.0~cxx+fortran+hl~ipo~java~map+mpi+shared~subfiling~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64
```
Note the `hdf5@1.14.3%gcc@11.4.0` above. This breaks dependent pkgs that attempt to `use hdf5` as the f90 module is no-longer compatible [with ifc].

With the snapshot prior to this change - I get:
```
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack.a$ ./bin/spack spec xsdk@1.0.0 %oneapi@2024.0.2 ^hdf5 %oneapi@2024.0.2 |grep hdf5@
 -           ^hdf5@1.14.3%oneapi@2024.0.2~cxx+fortran+hl~ipo~java~map+mpi+shared~subfiling~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64

```

Any idea if hdf5 needs a fix to restore this behavior?

Alternative I'm having to use now: 
```
svcpetsc@petsc-gpu-02:/scratch/svcpetsc/spack.a$ ./bin/spack spec xsdk@1.0.0 %oneapi@2024.0.2 ^hdf5 %oneapi@2024.0.2 |grep hdf5@
 -           ^hdf5@1.14.3%oneapi@2024.0.2~cxx+fortran+hl~ipo~java~map+mpi+shared~subfiling~szip~threadsafe+tools api=default build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64
```

alalazo commented:
@balay I'll check if I can reproduce this default behavior, and why it happens. Thanks for reporting.

mabraham commented:
I'm trying to make the GROMACS spack package get `sycl` from the provision here, but I see that nothing actually uses `IntelOneapiRuntime`. Is there a missing piece still? @alalazo

mabraham commented:
Using https://github.com/mabraham/spack/tree/mja-gromacs-sycl-from-oneapi, I observe
```
$ spack spec gromacs@git.intel-main=2025 %oneapi +sycl ^intel-oneapi-runtime@2025.0
==> Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    gromacs@git.intel-main=2025%oneapi+sycl ^intel-oneapi-runtime@2025.0 is unsatisfiable
```

mabraham commented:
Ah, working now: https://github.com/spack/spack/pull/47806
